sopel/modules/emailcheck.py

```diff
@@ -0,0 +1,314 @@
+
+def malicious_response(bot, nick: str, email):
+    fdrop(bot, nick)
+    add_badmail(bot, email)
+    bot.say(f"You have been temporarily banned from this network because {email.domain} "
+             "has a history of spam or abuse, and/or is a disposable email domain. "
+             "If this is a legitimate domain, contact staff for assistance.",
+             nick.lower())
+    gline_or_kill(bot, nick, bot.config.emailcheck.gline_time)
+    protect_chans(bot)
+    alert(bot, f"ALERT: User {nick} attempted to register a nick with disposable/spam domain {email.domain}!")
+
+def disallow_response(bot, nick: str, email):
+    fdrop(bot, nick)
+    add_badmail(bot, email)
+    bot.say(f"Your registration has been disallowed because {email.domain} appears to be suspicious. "
+             "If this is a legitimate domain, contact staff for assistance.",
+             nick.lower())
+    alert(bot, f"WARNING: User {nick} attempted to register a nick with suspicious domain {email.domain}.")
+
+def fetch_IPQS_email_score(
+    email_addr: str,
+    key: str,
+    fast: bool = True
+    ) -> tuple[float, bool, bool]: #score, disposable, has recent abuse flag set
+    '''Perform lookup on a specific email adress using ipqualityscore.com'''
+    email_str = urllib.parse.quote(email_addr)
+    faststr = str(bool(fast)).lower() #lower + handle None and other garbage
+    params = urllib.parse.urlencode({'fast': faststr})
+    with urllib.request.urlopen(
+        f"https://ipqualityscore.com/api/json/email/{key}/{email_str}?{params}") as url:
+            data = json.loads(url.read().decode())
+            LOGGER.debug(data)
+    if not data['success']:
+        errstr = f"{email_addr} lookup failed with {data['message']}"
+        LOGGER.error(errstr)
+        raise RuntimeError(errstr)
+    return (data['fraud_score'], data["disposable"], data["recent_abuse"])
+
+def get_email_score_from_db(session, email):
+    query_result = session.query(KnownEmails)\
+        .filter(KnownEmails.domain == email.domain)\
+        .one_or_none()
+    if query_result:
+        #Any known problematic provider should've been BADMAILed by now, but...
+        return DomainInfo(query_result.score,
+                          query_result.flag_disposable,
+                          query_result.flag_recent_abuse)
+
+def store_email_score_in_db(session, email, nick, IPQSresult):
+    new_known_email = KnownEmails(domain= email.doman[:DOMAIN_LEN],
+                                    first_nick= nick,
+                                    score= IPQSresult[0],
+                                    flag_disposable= IPQSresult[1],
+                                    flag_recent_abuse= IPQSresult[2])
+    session.add(new_known_email)
+    session.commit()
+
+def retrieve_score(bot, email, nick):
+    session = bot.db.session()
+    try:
+        global db_populated
+        if not db_populated:
+            BASE.metadata.create_all(bot.db.engine)
```

Comment:
This is kind of a weird place to initialize database tables

Reply:
Yup. Problem is that it has to be somewhere where the bot variable is available. I think setup() will work; I'll change that.